### ISSUE TITLE
Changing auto generated id in ingest pipeline to existing id

### ISSUE THREAD

**Issue by imotov** [elastic/elasticsearch#57693]
It is possible to change auto-generated id in an ingest pipeline processor, but if the resulted id already  exists in the index, the operation fails with `version conflict, document already exists` error.

To reproduce:

```
DELETE my_index

PUT my_index
{
  
}

PUT _ingest/pipeline/singleton
{
  "description": "Assigns the same id to all documents",
  "processors": [
    {
      "set": {
        "field": "_id",
        "value": "singleton"
      }
    }
  ]
}

POST my_index/_doc/?pipeline=singleton
{
  "val": 1
}

POST my_index/_doc/?pipeline=singleton
{
  "val": 2
}

GET my_index/_doc/singleton
```

Please note, that this issues is somewhat related to #41163, but it is not about accessing the id, but replacing id instead.

I am not sure if this is a bug or not, but it seems to be somewhat inconsistent. I feel like we shouldn't allow id modification in this scenario or we shouldn't make assumption about id uniqueness if we allow id to be changed. Personally I would prefer the latter.  

Background: I stumbled upon this when I was trying to force "Write to Index" connector in our new Kibana-based alerting system to update the same record. At the moment, Alerting in Kibana can only add records with auto-generated ids and they are using POST index/_doc to index all documents.

**Comment by ywelsch** [comment]
I'm not sure why you consider this a bug. You're adding a document using `POST` (which has the "op_type" default to "create", as documented [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html)), and then adapt the document ID to a given value, so the final indexing request is going to be executed as "op_type" : "create" with the given ID.

I'm relabeling this issue as "ingest", as the behavior here is determined by the ingest processor, not the core CRUD APIs.

**Comment by imotov** [comment]
Indeed, maybe an answer is as simple as exposing op in an ingest processor similar to update processor. 

Mentioning an `op_type` in the error message would have been helpful as well. I am just trying to retrace steps that an end user would have to make to come to this conclusion and we have a few issues here. It starts with Kibana not mentioning anywhere that `POST /index/_doc` format is used (had to figure it out from the behavior). The `op_type` behavior is indeed documented, but it is difficult to find unless you search for `op_type`, which bring us back to the original point that the error message doesn't mention `op_type` at all, so a user would have no idea that this is what they should be looking for:

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "version_conflict_engine_exception",
        "reason" : "[singleton]: version conflict, document already exists (current version [1])",
        "index_uuid" : "BW6zs4dYQfCgwL0OKLU6fg",
        "shard" : "0",
        "index" : "my_index"
      }
    ],
    "type" : "version_conflict_engine_exception",
    "reason" : "[singleton]: version conflict, document already exists (current version [1])",
    "index_uuid" : "BW6zs4dYQfCgwL0OKLU6fg",
    "shard" : "0",
    "index" : "my_index"
  },
  "status" : 409
}
```

I guess I consider it a bug because I don't see a way for an end user to figure it out. It might not be your bug, but it feels like our (stack) bug to me.

**Comment by ianadeem** [comment]
We would like to be able to use the ingest pipeline to update existing documents, too. I think we have a genuine use-case for this which I'll explain below.

We have a service S that exposes an API to fetch a bunch of metrics. The API returns the metrics cumulatively, i.e. totals since the beginning of time, on a per-user basis. Now I know it's suboptimal that the data is already pre-aggregated like this, but we have no control over S or its API. Since our company does all its analytics in Kibana, we want to ingest these metrics to Elasticsearch using a process P that sits in the background and pulls metrics from S once a day and ingests them using a pipeline. The pipeline is a must -- all our data is ingested through a common pipeline (and for good reasons it is not possible to change this behaviour).

P is stateless -- it doesn't know what's in Elasticsearch. It only needs to pull data from S and push it to Elasticsearch (one doc per user). But since the data is cumulative, we cannot just add new docs -- we must "upsert" the docs. To do so, we can set the document _id to the user's id in the pipeline -- but this issue stops us from doing so.

(The set of users is bounded and not too big -- re-ingesting data on a daily basis is not a problem).

**Comment by henningandersen** [comment]
@ianadeem , I believe you can fulfill that using the [`_update` API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html) (or a bulk request with updates).

**Comment by ianadeem** [comment]
> @ianadeem , I believe you can fulfill that using the [`_update` API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html) (or a bulk request with updates).

Thanks for the reply - I think I got confused here. In our case, we are actually using Filebeat for data ingestion with a simple TCP input. Filebeat sends the event to Elasticsearch with an ingest pipeline. I thought that this would always result in an `_index` API request, but I just stumbled upon the concept of de-duplication in Beats. Will do some more research!

**Comment by dagwieers** [comment]
We have a similar need for our use-case, we tend to send a database log to Elastic using the sql module, but also need to update existing records in some cases. Being able to set the `op_type` as part of the ingest pipeline would actually make this possible.
